In [43]:
import sys
import os

In [44]:
project_dir = os.path.dirname(os.getcwd())

In [45]:
sys.path.append(os.path.abspath(os.path.join(os.path.dirname("classifier.ipynb"), "..")))

## Import the Classes

In [46]:
from models.decisiontreeregressor import DecisionTreeRegressor
from models.randomforestregressor import RandomForestRegressor
from models.xgboostregressor import XGBoostRegressor
from models.linearregression import LinearRegression
from models.linearregression_closed import LinearRegressionClosed

In [6]:
import pandas as pd
import numpy as np

## Load the data

In [7]:
insurance = pd.read_csv(os.path.join(project_dir, "data/insurance.csv"))

In [8]:
insurance = pd.get_dummies(insurance, columns=["region"], dtype=int)
insurance["sex"] = insurance["sex"].map({"male": 1, "female": 0})
insurance["smoker"] = insurance["smoker"].map({"yes": 1, "no": 0})

In [9]:
X = insurance.drop(columns="charges")
y = insurance["charges"]

In [32]:
n = len(X)

In [33]:
indices = np.arange(n)
np.random.shuffle(indices)

In [34]:
s1 = int(0.75 * n)

In [35]:
X_train = X.iloc[indices[:s1],]
y_train = y.iloc[indices[:s1],]
X_test = X.iloc[indices[s1:],]
y_test = y.iloc[indices[s1:],]

## Linear Regression

In [47]:
import importlib
import models.linearregression

# Reload the module
importlib.reload(models.linearregression)

<module 'models.linearregression' from '/Users/vallurileelasaikrishna/Documents/from-scratch/models/linearregression.py'>

In [69]:
lr = LinearRegression(alpha=1e-30)

In [70]:
lr.fit(X_train, y_train)

In [74]:
lr.r_score(X_test, y_test)

np.float64(-1.223480186511976)

In [75]:
lr.rmse(X_test, y_test)

np.float64(18172.22960598861)

## Linear Regression Closed

In [38]:
lrc = LinearRegressionClosed()

In [39]:
lrc.fit(X_train, y_train)

In [40]:
lrc.r_score(X_test, y_test)

np.float64(-6.259838995190132)

In [41]:
lrc.rmse(X_test, y_test)

np.float64(35321.667513285734)

## Decision Tree

In [33]:
dtr = DecisionTreeRegressor()

In [34]:
dtr.fit(X_train, y_train)

In [35]:
dtr.rmse(X_test, y_test)

np.float64(4818.767123109063)

In [36]:
dtr.r_score(X_test, y_test)

np.float64(0.864880985129607)

## Random Forest

In [21]:
rfr = RandomForestRegressor()

In [22]:
rfr.fit(X_train, y_train)

In [23]:
rfr.rmse(X_test, y_test)

np.float64(5471.663064951269)

In [24]:
rfr.r_score(X_test, y_test)

np.float64(0.8228688837456295)

## XGBoost

In [25]:
xgbr = XGBoostRegressor()

xgb takes train and val at the same time for fit

In [26]:
s1 = int(0.7 * n)
s2 = int(0.85 * n)

In [27]:
X_train = X.iloc[indices[:s1],]
y_train = y.iloc[indices[:s1],]
X_val = X.iloc[indices[s1:s2],]
y_val = y.iloc[indices[s1:s2],]
X_test = X.iloc[indices[s2:],]
y_test = y.iloc[indices[s2:],]

In [29]:
xgbr.fit(X_train, y_train, X_val, y_val)

Early stopping at tree 289


In [31]:
xgbr.rmse(X_test, y_test)

np.float64(4803.722256735814)

In [32]:
xgbr.r_score(X_test, y_test)

np.float64(0.8657233890283251)

### Model Performance Comparison

| Model | RMSE | R² |
|------|--------:|-------:|
| Linear Regression | 18172.22960598861 | -1.223480186511976 | 
| Linear Regression Closed | 35321.667513285734 | -6.259838995190132 | 
| Decision Tree | 4818.767123109063 | 0.864880985129607 | 
| Random Forest | 5471.663064951269 | 0.8228688837456295 | 
| XGBoost | 4803.722256735814 | 0.8657233890283251 | 